File: docking/hw_testing/drone.py
# standard python library imports
import asyncio
import time
import math
import numpy as np

# mavsdk
from mavsdk import System
from mavsdk.offboard import (OffboardError, VelocityBodyYawspeed, VelocityNedYaw)

# aviata modules
import util
from image_analyzer import ImageAnalyzer
from pid_controller import PIDController
from log import Log

boom_length = 1 # meter
fov = 48.8
altitude_disp = boom_length / 2 / math.tan(math.radians(fov / 2)) * 2 # add 100% tolerance

class Drone:

    def __init__(self, target):
        # self.camera_simulator = CameraSimulator(target) # later, use camera here
        self.image_analyzer = ImageAnalyzer()
        self.north = 0
        self.east = 0
        self.down = 0
        self.yaw = 0
        self.takeoff_altitude = 2 # initial altitude
        self.target = target 
        self.dt = 0.05 # delta t update frequency, 20hz
        self.MAX_ATTEMPTS = 3 # number of unsuccessful attempts before aborting docking
        self.MAX_HEIGHT = 10 # max height above central target before failure
        self.MAX_HEIGHT_STAGE_2 = 2
        self.STAGE_1_TOLERANCE = 0.1 
        self.STAGE_2_TOLERANCE = 0.05
        self.log=Log()
        self.logging=True

        self.in_air = False

    async def connect_gazebo(self):
        """
        Connect to gazebo and perform basic setup checks.
        Does not return until connection to drone is verified.
        """
        # Wait for connection
        self.drone = System()
        await self.drone.connect(system_address="udp://:14540")

        async for state in self.drone.core.connection_state():
            if state.is_connected:
                self.log.write_literal(f"Connected to drone with UUID: {state.uuid}")
                break

        # Basic setup checks
        async for global_lock in self.drone.telemetry.health():
            if global_lock.is_global_position_ok:
                self.log.write_literal("Global position state is ok")
                break

    async def takeoff(self, test_number):
        """
        Takeoff and start offboard mode, then initiate docking.
        Return true if success, false otherwise
        """
        self.log.set_tagline("Takeoff")
        await self.drone.action.arm()

        # Get telemetry periodically from the drone
        telemetry_task = asyncio.create_task(self.get_telemetry_position())
        rotation_task = asyncio.create_task(self.get_telemetry_rotation())
        in_air_task = asyncio.create_task(self.get_telemetry_in_air())

        # Takeoff the vehicle (later, we can assume that the drone is already flying when we run this script)
        self.log.write_literal("-- Taking off")
        await self.drone.action.set_takeoff_altitude(2.0) # meters
        await self.drone.action.takeoff()
        
        while not self.in_air:
            await asyncio.sleep(0.1)
        
        await asyncio.sleep(8)

        # start offboard mode (requires setting initial setpoint)
        self.log.write_literal("Setting initial setpoint")
        await self.drone.offboard.set_velocity_body(VelocityBodyYawspeed(0.0, 0.0, 0.0, 0))

        self.log.write_literal("Starting offboard")
        try:
            await self.drone.offboard.start()
        except OffboardError as error:
            self.log.write_literal(f"Starting offboard mode failed with error code: {error._result.result}")
            self.log.write_literal("Disarming")
            await self.drone.action.disarm()
            return False

        self.log.write_literal(f"Takeoff complete, ascended to: {abs(self.down)}")

        # do tests here
        if test_number == 1:
            self.test1()
        elif test_number == 2:
            self.test2()

        await asyncio.sleep(3)
        await self.land()

        telemetry_task.cancel()
        rotation_task.cancel()

        return True

    def test1(self):
        # don't forget to set async if necessary
        self.log.set_tagline("Test 1")
        self.log.write_literal("We are in test 1")

    def test2(self):
        # don't forget to set async if necessary
        self.log.set_tagline("Test 2")
        self.log.write_literal("We are in test 2")

    async def get_telemetry_rotation(self):
        """Poll to obtain the drone's yaw"""
        await self.drone.telemetry.set_rate_attitude(20) # update at 20Hz 
        async for angles in self.drone.telemetry.attitude_euler():
            self.yaw = angles.yaw_deg # 0 degrees is north, positive cw, negative ccw, between -180 and +180


    async def get_telemetry_position(self):
        """Poll to obtain the drone's x, y, and z position"""
        await self.drone.telemetry.set_rate_position_velocity_ned(20) # update at 20Hz 
        async for position in self.drone.telemetry.position_velocity_ned():
            p = position.position
            self.north = p.north_m
            self.east = p.east_m
            self.down = p.down_m

    async def get_telemetry_in_air(self):
        async for is_in_air in self.drone.telemetry.in_air():
            self.in_air = is_in_air

    async def land(self):
        """Safely stop the drone and stop offboard mode"""
        self.log.set_tagline("Landing")
        self.log.write_literal("Stopping offboard")
        try:
            await self.drone.offboard.stop()
        except OffboardError as error:
            self.log.write_literal(f"Stopping offboard mode failed with error code: \
                {error._result.result}")

        self.log.write_literal("-- Landing drone")
        await self.drone.action.land()

File: docking/hw_testing/cpp/src/drone.cpp
#include "drone.hpp"
#include "util.hpp"

#include <iostream>
#include <future>
#include <string>

#include <mavsdk/mavsdk.h>
#include <mavsdk/plugins/action/action.h>
#include <mavsdk/plugins/offboard/offboard.h>
#include <mavsdk/plugins/telemetry/telemetry.h>

using namespace mavsdk;
using namespace cv;
using namespace std::chrono;
using std::this_thread::sleep_for;

Drone::Drone(Target t) 
    : m_north(0), m_east(0), m_down(-5), m_yaw(0), m_dt(0.05), m_target_info(t) //,image_analyzer()
{}

/**
 * Connect to Gazebo and run pre-flight checks
 * This function should be called before attempting to take any other action with the drone.
 * 
 * @return true if connection succeeded, false otherwise
 * */
bool Drone::connect_gazebo() {
    std::string tag = "Connecting";
    std::string connection_url = "udp://:14540"; // change as needed, or add command-line parsing
    ConnectionResult connection_result = mavsdk.add_any_connection(connection_url);

    if (connection_result != ConnectionResult::Success) {
        log(tag, "Connection failed, see next line", true);
        std::cout << connection_result << std::endl;
        return false;
    }

    log(tag, "Waiting to discover system");
    std::promise<void> discover_promise;
    auto discover_future = discover_promise.get_future();

    mavsdk.subscribe_on_new_system([&mavsdk=mavsdk, &discover_promise, &tag]() {
        const auto system = mavsdk.systems().at(0);

        if (system->is_connected()) {
            log(tag, "Discovered system");
            discover_promise.set_value();
        }
    });

    discover_future.wait();

    // System got discovered.
    m_system = mavsdk.systems().at(0);
    auto telemetry = Telemetry{m_system};

    while (!telemetry.health_all_ok()) {
        log(tag, "Waiting for system to be ready");
        sleep_for(seconds(1));
    }

    log(tag, "System is ready");
    return true;
}

/**
 * Attempt to take off.
 * 
 * @return true if successful, false otherwise
 * */
bool Drone::takeoff() {
    std::string tag = "Takeoff";
    std::promise<void> in_air_promise;
    auto in_air_future = in_air_promise.get_future();

    // Arm the drone
    auto action = Action{m_system};
    Action::Result arm_result = action.arm();
    if (arm_result != Action::Result::Success) {
        log(tag, "Arming failed", true);
        return false;
    }
    log(tag, "Armed");

    // Attempt to take off
    action.set_takeoff_altitude(8);
    Action::Result takeoff_result = action.takeoff();
    if (takeoff_result != Action::Result::Success) {
        log(tag, "Takeoff failed, error code on next line", true);
        std::cout << "Code: " << takeoff_result << std::endl;
        return false;
    }

    log(tag, "Taking off");

    // Wait for the take off to be complete
    auto telemetry = Telemetry{m_system};
    telemetry.subscribe_landed_state([&in_air_promise, &telemetry, &tag](Telemetry::LandedState landed) {
        if (landed == Telemetry::LandedState::InAir) {
            telemetry.subscribe_landed_state(nullptr);
            in_air_promise.set_value();
            log(tag, "Take off successful");
        }
    });

    in_air_future.wait(); // wait for the asynchronous call to be finished (when in_air_promise resolves)

    // Start offboard mode
    auto offboard = Offboard{m_system};
    Offboard::VelocityNedYaw stay{};
    offboard.set_velocity_ned(stay); // set initial setpoint 

    Offboard::Result offboard_result = offboard.start();
    if (offboard_result != Offboard::Result::Success) {
        log(tag, "Offboard start failed, error code on next line", true);
        std::cout << "Code: " << offboard_result << std::endl;
        return false;
    }

    log(tag, "Offboard started");

    return true;
}

void Drone::initiate_docking(int target_id) {

    bool stage1_outcome = stage1(target_id);
    if (stage1_outcome) {
        stage2(target_id);

    }   
}

void Drone::set_position(float north, float east, float down) {
    m_north = north;
    m_east = east;
    m_down = down;
}

void Drone::set_yaw(float yaw) {
    m_yaw = yaw;
}

bool Drone::stage1(int target_id) {


    // Register for attitude updates
    auto telemetry = Telemetry{m_system}; // NOTE: telemetry updates are only received as long as the telemetry pointer is alive
                                          // will need to either resubscribe in stage 2 or move this to initiate_docking to keep it alive
                                          // and pass as argument to stage1(), stage2()
    const Telemetry::Result set_rate_result_p = telemetry.set_rate_position_velocity_ned(40);
    const Telemetry::Result set_rate_result_a = telemetry.set_rate_attitude(40);
    if (set_rate_result_a != Telemetry::Result::Success || set_rate_result_p != Telemetry::Result::Success) {
        std::cout << "Setting attitude rate possibly failed:" << set_rate_result_a << '\n';
        std::cout << "Setting position rate possibly failed:" << set_rate_result_p << '\n';
    }

    // Subscribe to getting periodic telemetry updates with lambda function
    telemetry.subscribe_position_velocity_ned([this](Telemetry::PositionVelocityNed p) {
        set_position(p.position.north_m, p.position.east_m, p.position.down_m);
    });
    telemetry.subscribe_attitude_euler([this](Telemetry::EulerAngle e) {
        set_yaw(e.yaw_deg);
    });

    auto offboard = Offboard{m_system};

    while (true) {
        high_resolution_clock::time_point t1 = high_resolution_clock::now();

        Offboard::VelocityBodyYawspeed spin{};
        spin.yawspeed_deg_s = 36.0f;
        offboard.set_velocity_body(spin);
        log("Yaw", std::to_string(m_yaw));
        
        Mat currImg = raspi_camera.update_current_image(); // looks a bit choppy only cause it's at 20fps, remove sleep_for and it's smooth
        // string tags="";
        // float* errs=image_analyzer.processImage(currImg,0,m_yaw,tags);
        // log("Tags",tags);

        high_resolution_clock::time_point t2 = high_resolution_clock::now();
        duration<double, std::milli> d = (t2 - t1);
        int time_span = d.count();
        if (time_span < m_dt * 1000) {
            sleep_for(milliseconds((int) (m_dt * 1000 - time_span)));
        }
    }
    return true;
}

void Drone::stage2(int target_id) {

}

void Drone::offset_errors(double x, double y, double alt, double rot, int target_id) {

}

void Drone::safe_land() {
    land();
}

void Drone::land() {

}